pkg/model/series.go

```diff
@@ -70,15 +103,141 @@ func (m *SeriesMerger) mergePoints(points []*typesv1.Point) int {
 		return points[i].Timestamp < points[j].Timestamp
 	})
 	var j int
-	for i := 1; i < l; i++ {
-		if points[j].Timestamp != points[i].Timestamp {
-			j++
-			points[j] = points[i]
+	for i := 0; i < l; i++ {
+		if m.aggregator.IsEmpty() {
+			m.aggregator.Add(points[i].Timestamp, points[i].Value)
 			continue
 		}
-		if m.sum {
-			points[j].Value += points[i].Value
+		if m.aggregator.GetTimestamp() != points[i].Timestamp {
+			points[j] = m.aggregator.GetAndReset()
+			j++
+			m.aggregator.Add(points[i].Timestamp, points[i].Value)
+			continue
 		}
+		m.aggregator.Add(points[i].Timestamp, points[i].Value)
+	}
+	if !m.aggregator.IsEmpty() {
+		points[j] = m.aggregator.GetAndReset()
 	}
 	return j + 1
 }
+
+type TimeSeriesAggregator interface {
+	Add(ts int64, value float64)
+	GetAndReset() *typesv1.Point
+	IsEmpty() bool
+	GetTimestamp() int64
+}
+
+func NewTimeSeriesAggregator(aggregation *string) TimeSeriesAggregator {
+	if aggregation == nil || *aggregation == sumAggregationType {
+		return &sumTimeSeriesAggregator{
+			ts: -1,
+		}
+	}
+	if *aggregation == AverageAggregationType {
+		return &avgTimeSeriesAggregator{
+			ts: -1,
+		}
+	} else if *aggregation == firstValueAggregationType {
+		return &firstValueTimeSeriesAggregator{
+			ts: -1,
+		}
+	} else {
+		return &sumTimeSeriesAggregator{
+			ts: -1,
+		}
+	}
+}
+
+type sumTimeSeriesAggregator struct {
+	ts  int64
+	sum float64
+}
+
+func (a *sumTimeSeriesAggregator) Add(ts int64, value float64) {
+	a.ts = ts
+	a.sum += value
+}
+
+func (a *sumTimeSeriesAggregator) GetAndReset() *typesv1.Point {
+	tsCopy := a.ts
+	sumCopy := a.sum
+	a.ts = -1
+	a.sum = 0
+	return &typesv1.Point{
+		Timestamp: tsCopy,
+		Value:     sumCopy,
+	}
+}
+
+func (a *sumTimeSeriesAggregator) IsEmpty() bool {
+	return a.ts == -1
+}
+
+func (a *sumTimeSeriesAggregator) GetTimestamp() int64 {
+	return a.ts
+}
+
+type avgTimeSeriesAggregator struct {
+	ts    int64
+	sum   float64
+	count int64
+}
+
+func (a *avgTimeSeriesAggregator) Add(ts int64, value float64) {
+	a.ts = ts
+	a.sum += value
+	a.count++
+}
```

Comment:
I'd like to make sure I fully understand the aggregation and its relation to grouping. I'm sorry if I got it wrong :) 

Given two series (with just one value each):
```
{pod=a, endpoint=x}: { ts_1, 8 }
{pod=a, endpoint=y}: { ts_1, 4 }
```

Given a query `avg (group) by pod`. What would be the result: `{pod=a} 12` or `{pod=a} 6`?

My understanding is that the aggregator gets two values at input (from **[seriesBuilder](https://github.com/grafana/pyroscope/blob/main/pkg/phlaredb/sample_merge.go#L200)**):
```
{pod=a}: { ts_1, 4 }, { ts_1, 8 }
```
Which then will be averaged and the result will be 6. I'm not very sure this is what users expect (despite it is literally the average value)

Reply:
It will be `{pod=a} 6` and I am not sure either though summing will often make no sense as well, depending on the profile type. There are two aspects to this aggregation:

1. Reducing samples for the exact same timestamp (happens in [series.go](https://github.com/grafana/pyroscope/blob/94d0a6a03c1cc895a575c36e8adb162891f938f6/pkg/model/series.go#L79-L81), previously either "first" or "sum")
2. Reducing samples within a step interval (happens in [querier.go](https://github.com/grafana/pyroscope/blob/94d0a6a03c1cc895a575c36e8adb162891f938f6/pkg/querier/querier.go#L972), previously always "sum")

Honestly I am not sure we need 1. Leaving duplicate samples in and letting 2 handle the aggregation would produce more reliable results (currently in 2 we will do avg of avg at times).

Reply:
Totally agree we should only do 2 to avoid avg of avg. May be only when it's not a sum.